File: leetcode/0052-n-queens-ii/Solution.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    private int count;
    private final Set<Integer> usedCols = new HashSet<>();
    private final Set<Integer> usedDiag1 = new HashSet<>();
    private final Set<Integer> usedDiag2 = new HashSet<>();

    public int totalNQueens(int n) {
        count = 0;
        backtrack(0, n);
        return count;
    }

    private void backtrack(int row, int n) {
        if (row == n) {
            count++;
            return;
        }

        for (int col = 0; col < n; col++) {
            int diag1 = row - col;
            int diag2 = row + col;
            if (usedCols.contains(col) || usedDiag1.contains(diag1) || usedDiag2.contains(diag2)) {
                continue;
            }

            usedCols.add(col);
            usedDiag1.add(diag1);
            usedDiag2.add(diag2);

            backtrack(row + 1, n);

            usedCols.remove(col);
            usedDiag1.remove(diag1);
            usedDiag2.remove(diag2);
        }
    }
}

File: leetcode/0052-n-queens-ii/solution.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        self.count = 0
        used_cols = set()
        used_diag1 = set()  # row - col
        used_diag2 = set()  # row + col

        def backtrack(row: int) -> None:
            if row == n:
                self.count += 1
                return

            for col in range(n):
                if col in used_cols or (row - col) in used_diag1 or (row + col) in used_diag2:
                    continue

                used_cols.add(col)
                used_diag1.add(row - col)
                used_diag2.add(row + col)

                backtrack(row + 1)

                used_cols.remove(col)
                used_diag1.remove(row - col)
                used_diag2.remove(row + col)

        backtrack(0)
        return self.count
